From f1ea775d9c6de9bec8372ce9d305f60c7546a54c Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Fri, 4 Apr 2025 11:39:38 +0000
Subject: [PATCH] fix musl build with linux/prctl.h

musl libc's sys/prctl.h lacks PR_SET_VMA but has conflicting
definitions with linux/prctl.h, causing struct redefinition
errors during build:

    In file included from /ws/workspace/aq-build-pkg/aports/core/mimalloc/src/mimalloc-2.2.3/src/prim/unix/prim.c:36,
                     from /ws/workspace/aq-build-pkg/aports/core/mimalloc/src/mimalloc-2.2.3/src/prim/prim.c:25:
    /usr/include/sys/prctl.h:88:8: error: redefinition of 'struct prctl_mm_map'
       88 | struct prctl_mm_map {
          |        ^~~~~~~~~~~~
    In file included from /ws/workspace/aq-build-pkg/aports/core/mimalloc/src/mimalloc-2.2.3/src/prim/unix/prim.c:34:
    /usr/include/linux/prctl.h:134:8: note: originally defined here
      134 | struct prctl_mm_map {
          |        ^~~~~~~~~~~~

Manually define PR_SET_VMA as a temporary workaround until musl's
headers are updated.
---
 src/prim/unix/prim.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/prim/unix/prim.c b/src/prim/unix/prim.c
index 780d254f..710515aa 100644
--- a/src/prim/unix/prim.c
+++ b/src/prim/unix/prim.c
@@ -35,6 +35,10 @@ terms of the MIT license. A copy of the license can be found in the file
   #if defined(__GLIBC__) && !defined(PR_SET_VMA)
   #include <linux/prctl.h>
   #endif
+# ifndef PR_SET_VMA
+#   define PR_SET_VMA		0x53564d41
+#   define PR_SET_VMA_ANON_NAME		0
+# endif
   #if defined(__GLIBC__)
   #include <linux/mman.h>   // linux mmap flags
   #else
-- 
2.25.1

